README.md
Marello Api Extension
=============

Marello Api extension is a wrapper which enables the applications to communicate with an
instance of Marello. This wrapper only provides authentication with X-WSSE headers and allows you to send or get data from a Marello Application instance.

The extension is based on the work Sylvain Rayé has done back in 2014, where he explains how to connect to an OroCRM/Akeneo instance with this method.
The extension uses a lot of work done by Sylvain Rayé regarding the X-WSSe authentication but adds a Client into it to make the create and send the calls to an instance.
The original post can found  on [Sylvain Rayé's blog](http://www.sylvainraye.com/2014/03/23/using-the-rest-api-of-oroplatform-orocrm-akeneo/).
Oro's cookbook now includes an [instruction](https://github.com/orocrm/documentation/blob/master/cookbook/how_to_use_wsse_authentication.rst) on how to generate the X-WSSE in an Oro based platform

## Features
- Pinging the Marello instance (\Marello\Api\Client::pingInstance());
- Getting data from Marello, including page/limit filter;
- Sending data to Marello with POST/PUT/DELETE requests

**Future Features**
- Disable denpendency from Marello Bridge on this wrapper for swapping out this wrapper for, for example GuzzleHttp

## Requirements

* PHP 5.5.0 or above with command line interface

## Installation instructions

In order to get the Bridge Api, you can easily install this through composer. If you don't have composer installed globally, you can get it by running the following command:
```bash
curl -s https://getcomposer.org/installer | php
```

```bash
php composer.phar require "marellocommerce/marello-bridge-api"
```

- Install dependencies with composer. If installation process seems too slow you can use `--prefer-dist` option.

```bash
php composer.phar install --prefer-dist --no-dev
```


## Running Tests
In order to run the tests for the Marello Bridge API, you need to update the dependencies with composer by running the following command:
```bash
php composer.phar update
```

Or do a fresh install
```bash
php composer.phar install --prefer-dist
```

- To run the tests you need to run the following command in the `vendor/marellocommerce/marello-bridge-api` directory
```bash
vendor/phpunit/phpunit/phpunit --testsuite="Marello Api Test Suite"
```

or if you have `phpunit` installed globally:
```bash
phpunit --testsuite="Marello Api Test Suite"
```

## Usage
for usage see [docs/USAGE.md](doc/USAGE.md)

## Contact
Questions? Problems? Improvements?

Feel free to contact us either through [http://www.marello.com/contact/](http://www.marello.com/contact/), forum [http://www.marello.com/forum/marello/](http://www.marello.com/forum/marello/) or open an issue in the repository :) Thanks!

CHANGELOG.md
CHANGELOG for 1.1.1
=================
This changelog references the relevant changes (new features, changes and bugs) done in 1.1.1 version.
  * fix issue regarding not correctly formatted url in GET requests with payload

CHANGELOG for 1.1.0
=================
This changelog references the relevant changes (new features, changes and bugs) done in 1.1.0 version.
  * Added response code, Response and RequestHeader
  * Fix issue with default filters when doing GET requests with payload
  * deprecated `pingUsers()` method in favor of `pingInstance()`
  
  
Response will contain all the information based of `curl_getinfo()` and the actual result from `curl_exec()`. 

RequestHeader will contain all the data that has been send with the Curl request, this includes the payload if available

tests/Marello/ClientTest.php
<?php

/**
 * Marello
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is published at http://opensource.org/licenses/osl-3.0.php.
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@marello.com so we can send you a copy immediately
 *
 * @category  Marello
 * @package   Tests
 * @copyright Copyright Marello (http://www.marello.com)
 * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
*/
namespace Marello\Tests;

use Marello\Api\Client;
use Marello\Tests\Stub\ClientStub;

class ClientTest extends \PHPUnit_Framework_TestCase
{
    const TEST_URL = 'http://demo.marello.com/';
    const USERNAME = 'admin';
    const PASSWORD = 'd07009855780a53f025f0cfea3eb081ba124c0c8';

    /** @var ClientStub $client */
    protected $client;

    /**
     * {@inheritdoc}
     */
    public function setUp()
    {
        $this->client = new ClientStub(self::TEST_URL);
        $this->client->setAuth([
            'username' => self::USERNAME,
            'api_key' => self::PASSWORD
        ]);
    }

    /**
     * {@inheritdoc}
     */
    public function tearDown()
    {
        $this->client = null;
    }

    /**
     * @expectedException \Marello\Api\MarelloException
     * @expectedExceptionMessage No api url provided
     */
    public function testClientFailedToSetupUrl()
    {
        $client = new Client(null);
        $client->pingInstance();
    }

    /**
     * @expectedException \Marello\Api\MarelloException
     * @expectedExceptionMessage No credentials provided. Please call Client::setAuth() before making requests.
     */
    public function testClientFailedToSetupAuth()
    {
        $client = new Client(self::TEST_URL);
        $client->pingInstance();
    }

    /**
     * {@inheritdoc}
     */
    public function testClientSuccesfulPingInstance()
    {
        $this->client->setFakeResponse([
            'http_code' => 200,
            'content_type' => 'application/json',
            'body' => json_encode(['response' => 'is OK!'])
        ]);

        $this->assertTrue($this->client->pingInstance());
        $this->assertEquals(Client::HTTP_CODE_OK, $this->client->getResponseCode());
        $this->assertNotEmpty($this->client->getLastResponse());
    }

    /**
     * {@inheritdoc}
     */
    public function testRestGetCallWithoutQuery()
    {
        $this->client->setFakeResponse([
            'http_code' => 200,
            'content_type' => 'application/json',
            'body' => json_encode(['response' => 'is OK!'])
        ]);

        $result = $this->client->restGet('/users');
        $this->doLastResponseAndHeadersTests($result);
    }

    /**
     * {@inheritdoc}
     */
    public function testRestGetCallWithQuery()
    {
        $this->client->setFakeResponse([
            'http_code' => 200,
            'content_type' => 'application/json',
            'body' => json_encode(['response' => 'is OK!'])
        ]);

        $result = $this->client->restGet('/test', ['page' => 2, 'limit' => 200]);
        $this->doLastResponseAndHeadersTests($result);
        $this->assertContains('page=2&limit=200', $result->getData('url'));
    }

    /**
     * @expectedException \Marello\Api\MarelloException
     * @expectedExceptionMessage Query cannot be empty when a POST call is made.
     */
    public function testRestFailPostCall()
    {
        $this->client->restPost('/test', []);
    }

    /**
     * {@inheritdoc}
     */
    public function testRestPostCall()
    {
        $payload = ['customer' => 2, 'name' => 'john'];
        $result = $this->client->restPost('/test', $payload);

        $this->doLastResponseAndHeadersTests($result);
        $this->assertContains(json_encode($payload), $this->client->getLastRequestHeaders()->getData());
    }

    /**
     * @expectedException \Marello\Api\MarelloException
     * @expectedExceptionMessage Query cannot be empty when a PUT call is made.
     */
    public function testRestFailPutCall()
    {
        $this->client->restPut('/test', []);
    }

    /**
     * {@inheritdoc}
     */
    public function testRestPutCall()
    {
        $payload = ['order' => 2, 'lastName' => 'Doe'];
        $result = $this->client->restPost('/test', $payload);

        $this->doLastResponseAndHeadersTests($result);
        $this->assertContains(json_encode($payload), $this->client->getLastRequestHeaders()->getData());
    }

    /**
     * @expectedException \Marello\Api\MarelloException
     * @expectedExceptionMessage Query cannot be empty when a DELETE call is made.
     */
    public function testRestFailDeleteCall()
    {
        $this->client->restDelete('/test', []);
    }

    /**
     * {@inheritdoc}
     */
    public function testRestDeleteCall()
    {
        $payload = ['address' => 2];
        $result = $this->client->restPost('/test', $payload);

        $this->doLastResponseAndHeadersTests($result);
        $this->assertContains(json_encode($payload), $this->client->getLastRequestHeaders()->getData());
    }

    /**
     * {@inheritdoc}
     */
    public function testCallWithErrorResponse()
    {
        $this->client->setFakeResponse([
            'http_code' => 404,
            'content_type' => 'application/json',
        ]);
        $this->client->setResult(false);
        $errorMessage = 'Well this ain\'t good...';
        $this->client->setCurlError($errorMessage);

        $result = $this->client->restGet('/darn');
        $this->doLastResponseAndHeadersTests($result);
        $this->assertEquals(404, $this->client->getResponseCode());

        $this->assertArrayHasKey('error', $result->getData());
        $this->assertArrayNotHasKey('body', $result->getData());
        $this->assertEquals($errorMessage, $result->getErrorMessage());
    }

    /**
     * {@inheritdoc}
     * @param $result
     */
    protected function doLastResponseAndHeadersTests($result)
    {
        $this->assertNotEmpty($this->client->getLastResponse());
        $this->assertEquals($this->client->getLastResponse(), $result);
        $this->assertContains(self::TEST_URL, $this->client->getLastRequestHeaders()->getData());
    }
}


tests/Marello/Stub/ClientStub.php
<?php
/**
 * Marello
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is published at http://opensource.org/licenses/osl-3.0.php.
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@marello.com so we can send you a copy immediately
 *
 * @category  Marello
 * @package   Api
 * @copyright Copyright 2016 Marello (http://www.marello.com)
 * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 */
namespace Marello\Tests\Stub;

use Marello\Api\Client;
use Marello\Api\MarelloException;

class ClientStub extends Client
{
    /** @var string $requestUrl */
    protected $requestUrl;

    /** @var array $curlResponse */
    protected $curlResponse;

    /** @var string $curlError */
    protected $curlError;

    /** @var bool $result */
    protected $result = true;

    /**
     * Stub the callServer method to setup results and errors
     * @param string $url url to call
     * @param array|null $data data for the request
     * @param array $curlOptions
     * @throws MarelloException
     * @return mixed $result
     */
    public function callServer($url, $curlOptions, array $data = [])
    {
        $ch = curl_init($url);
        $request = $this->createRequest($ch, $curlOptions, $url);
        $this->setLastRequestHeaders($request);
        curl_close($ch);

        $result = $this->result; // normally the curl call is being executed

        $curlInfo = $this->curlResponse;
        if ($result === false) {
            $curlInfo['error'] = $this->curlError;
        } else {
            $curlInfo['body'] = $result;
            $curlInfo['url'] = $url;
        }

        $response = $this->createResponse($curlInfo);
        $this->setLastResponse($response);
        return $this->lastResponse;
    }

    /**
     * @param array $data
     */
    public function setFakeResponse(array $data)
    {
        $this->curlResponse = $data;
    }

    /**
     * @param $message
     */
    public function setCurlError($message)
    {
        $this->curlError = $message;
    }

    public function setResult($bool)
    {
        $this->result = $bool;
    }

    public function getApiUrl()
    {
        return $this->requestUrl;
    }
}


tests/Marello/AuthenticationTest.php
<?php

/**
 * Marello
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is published at http://opensource.org/licenses/osl-3.0.php.
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@marello.com so we can send you a copy immediately
 *
 * @category  Marello
 * @package   Tests
 * @copyright Copyright Marello (http://www.marello.com)
 * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 */
namespace Marello\Tests;

use Marello\Api\Authentication;
use Marello\Api\MarelloException;

class AuthenticationTest extends \PHPUnit_Framework_TestCase
{
    const TEST_USER_NAME    = 'admin';
    const TEST_PASSWORD     = '1234';

    /**
     * @expectedException \Marello\Api\MarelloException
     * @expectedExceptionMessage No Username or Apikey specified
     */
    public function testAuthenticationFailedToSetup()
    {
        $authentication = new Authentication(null, null);
        $authentication->getHeaders();
    }

    /**
     * {@inheritdoc}
     */
    public function testAuthenticationHeaders()
    {
        $authentication = new Authentication(self::TEST_USER_NAME, self::TEST_PASSWORD);
        $headers = $authentication->getHeaders();
        $this->assertCount(3, $headers);
        $this->assertTrue(is_array($headers));

        $this->assertContains('Content-Type: application/json', $headers);
        $this->assertContains('Authorization: WSSE profile="UsernameToken"', $headers);

        $lastHeader = end($headers);
        $this->assertContains('X-WSSE: UsernameToken', $lastHeader);
        $this->assertContains('Username=', $lastHeader);
        $this->assertContains('PasswordDigest=', $lastHeader);
        $this->assertContains('Nonce=', $lastHeader);
        $this->assertContains('Created=', $lastHeader);
    }
}


src/Marello/Api/Curl/Response.php
<?php

/**
 * Marello
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is published at http://opensource.org/licenses/osl-3.0.php.
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@marello.com so we can send you a copy immediately
 *
 * @category  Marello
 * @package   Api
 * @copyright Copyright 2016 Marello (http://www.marello.com)
 * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 */
namespace Marello\Api\Curl;

use Marello\Api\MarelloException;

class Response
{
    protected $responseData = [];

    /**
     * @param array $curlResult
     * @return Response
     * @throws MarelloException
     */
    public static function create(array $curlResult)
    {
        if (empty($curlResult)) {
            throw new MarelloException('Could not create responds from empty array');
        }

        $response = new self();
        $response->setData($curlResult);

        return $response;
    }

    public function getErrorMessage()
    {
        return $this->getData('error');
    }

    protected function setData($data)
    {
        $this->responseData = $data;
    }

    public function getData($key = null)
    {
        if (is_null($key)) {
            return $this->responseData;
        }

        if (!isset($this->responseData[$key]) || empty($this->responseData[$key])) {
            return null;
        }

        return $this->responseData[$key];
    }
}


src/Marello/Api/Authentication.php
<?php

/**
 * Marello
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is published at http://opensource.org/licenses/osl-3.0.php.
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@marello.com so we can send you a copy immediately
 *
 * @category  Marello
 * @package   Api
 * @copyright Copyright 2016 Marello (http://www.marello.com)
 * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 */
namespace Marello\Api;

use Marello\Api\MarelloException;

class Authentication
{
    /** @var string $username */
    private $username;

    /** @var string $apiKey */
    private $apiKey;

    /** @var array $headers */
    protected $headers;

    /**
     * Authentication constructor.
     * @param $username
     * @param $apiKey
     */
    public function __construct($username, $apiKey)
    {
        $this->setupAuthenticationCredentials($username, $apiKey);
    }

    /**
     * @param $username
     * @param $apiKey
     * @throws MarelloException
     */
    private function setupAuthenticationCredentials($username, $apiKey)
    {
        if (!$username || !$apiKey) {
            throw new MarelloException('No Username or Apikey specified');
        }
        $this->username = $username;
        $this->apiKey   = $apiKey;
    }

    /**
     * encode raw password
     * @param $raw
     * @return string
     */
    private function encodePassword($raw, $salt)
    {
        $salted = $this->mergePasswordAndSalt($raw, $salt);
        $digest = hash('sha1', $salted, true);
        return base64_encode($digest);
    }

    /**
     * @param $password
     * @param $salt
     * @return string
     * @throws \InvalidArgumentException
     */
    private function mergePasswordAndSalt($password, $salt)
    {
        if (empty($salt)) {
            return $password;
        }

        if (false !== strrpos($salt, '{') || false !== strrpos($salt, '}')) {
            throw new \InvalidArgumentException('Cannot use { or } in salt.');
        }

        return $password.'{'.$salt.'}';
    }

    /**
     * Prepare headers for the authentication
     * @return array
     */
    public function getHeaders()
    {
        $created = date('c');
        $nonce  = $this->generateNonce();
        $passwordDigest = $this->generatePasswordDigest($nonce, $created);

        $wsseProfile = sprintf(
            'X-WSSE: UsernameToken Username="%s", PasswordDigest="%s", Nonce="%s", Created="%s"',
            $this->username,
            $passwordDigest,
            $nonce,
            $created
        );

        return array(
            'Content-Type: application/json',
            'Authorization: WSSE profile="UsernameToken"',
            $wsseProfile
        );
    }

    /**
     * Generate nonce
     * @return string
     */
    private function generateNonce()
    {
        $prefix = gethostname();
        return base64_encode(substr(md5(uniqid($prefix . '_', true)), 0, 16));
    }

    /**
     * Generate password digest
     * @param $nonce
     * @param $created
     * @return string
     */
    private function generatePasswordDigest($nonce, $created)
    {
        $rawPass = base64_decode($nonce) . $created . $this->apiKey;
        return $this->encodePassword($rawPass, '');
    }
}


src/Marello/Api/Curl/RequestHeader.php
<?php

/**
 * Marello
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is published at http://opensource.org/licenses/osl-3.0.php.
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@marello.com so we can send you a copy immediately
 *
 * @category  Marello
 * @package   Api
 * @copyright Copyright 2016 Marello (http://www.marello.com)
 * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 */
namespace Marello\Api\Curl;

use Marello\Api\MarelloException;

class RequestHeader
{
    protected $headerData = [];

    /**
     * @param array $curlData
     * @return RequestHeader
     * @throws MarelloException
     */
    public static function create(array $curlData)
    {
        if (empty($curlData)) {
            throw new MarelloException('Could not create headers from empty array');
        }

        $header = new self();
        $header->setData($curlData);

        return $header;
    }

    protected function setData($data)
    {
        $this->headerData = $data;
    }

    public function getData($key = null)
    {
        if (!$key) {
            return $this->headerData;
        }

        if (!isset($this->headerData[$key]) || empty($this->headerData[$key])) {
            return null;
        }

        return $this->headerData[$key];
    }
}


src/Marello/Api/Client.php
<?php
/**
 * Marello
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is published at http://opensource.org/licenses/osl-3.0.php.
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@marello.com so we can send you a copy immediately
 *
 * @category  Marello
 * @package   Api
 * @copyright Copyright 2016 Marello (http://www.marello.com)
 * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 */
namespace Marello\Api;

use Marello\Api\Curl\RequestHeader;
use Marello\Api\Curl\Response;

class Client
{
    const HTTP_CODE_OK = 200;
    const CALL_PATH_REST_API = 'api/rest';

    /** @var string $version */
    protected $version = 'latest';

    /** @var array $credentials */
    protected $credentials = [];

    /** @var string $apiUrl */
    protected $apiUrl;

    /** @var int $responseCode */
    protected $responseCode;

    /** @var Response $lastResponse */
    protected $lastResponse;

    /** @var  $lastRequestHeaders */
    protected $lastRequestHeaders;

    /**
     * Marello_Api_Client constructor.
     * @param $apiUrl
     * @throws MarelloException
     */
    public function __construct($apiUrl)
    {
        if (!$apiUrl) {
            throw new MarelloException('No api url provided');
        }

        $this->setApiUrl($apiUrl);
    }

    /**
     * Set credentials for authentication
     * [
     *  'username' => 'username',
     *  'api_key' => 'apiKey',
     * ]
     *
     *
     * @param array $credentials
     * @throws MarelloException
     */
    public function setAuth(array $credentials)
    {
        if ((!isset($credentials['username']) && empty($credentials['username']))
            || (!isset($credentials['api_key']) && empty($credentials['api_key']))) {
            throw new MarelloException('Username of Apikey not specified.');
        }

        $this->credentials = $credentials;
    }

    /**
     * validate for a connection/authorization with Marello API
     * @deprecated use pingInstance() instead
     * @return bool
     */
    public function pingUsers()
    {
        return $this->pingInstance();
    }

    /**
     * validate for a connection/authorization with Marello API
     * @return bool
     */
    public function pingInstance()
    {
        $this->restGet("/users");

        if ($this->getResponseCode() !== self::HTTP_CODE_OK) {
            return false;
        }

        return true;
    }

    /**
     * Get response code from request
     * @return int
     */
    public function getResponseCode()
    {
        return $this->getLastResponse()->getData('http_code');
    }

    /**
     * Set api url including callpath and version
     * @param string $url
     */
    public function setApiUrl($url)
    {
        $this->apiUrl = $url . self::CALL_PATH_REST_API . DIRECTORY_SEPARATOR . $this->version;
    }

    /**
     * Prepare Rest uri for call
     * @param $path
     * @param $baseUri
     * @return string uri
     */
    private function prepareRestUri($path, $baseUri = null)
    {
        $uri = (is_null($baseUri)) ? $this->apiUrl : $baseUri;
        $skip = false;
        if (strpos($path, 'filter') !== false || strpos($path, 'filter') === 0) {
            $skip = true;
            $path = str_replace('filter', '', $path);
        }

        if ($path[0] != '/' && $uri[strlen($uri)-1] != '/' && !$skip) {
            $path = '/' . $path;
        }

        return $uri . $path;
    }

    /**
     * Set curl data and execute
     * @param string $url url to call
     * @param array|null $data data for the request
     * @param array $curlOptions
     * @throws MarelloException
     * @return mixed $result
     */
    public function callServer($url, $curlOptions, array $data = [])
    {
        $ch = curl_init($url);
        $request = $this->createRequest($ch, $curlOptions, $url);
        $this->setLastRequestHeaders($request);
        $result = curl_exec($ch);

        $curlInfo = curl_getinfo($ch);
        if ($result === false) {
            $curlInfo['error'] = curl_error($ch);
        } else {
            $curlInfo['body'] = $result;
        }

        curl_close($ch);

        $response = $this->createResponse($curlInfo);
        $this->setLastResponse($response);
        return $this->lastResponse;
    }

    /**
     * @param $info
     * @return Response
     */
    public function createResponse($info)
    {
        return Response::create($info);
    }

    /**
     * @param Response $response
     */
    public function setLastResponse(Response $response)
    {
        $this->lastResponse = $response;
    }

    /**
     * @param $handle
     * @param $data
     * @param $url
     * @return RequestHeader $curlData
     */
    public function createRequest($handle, $data, $url)
    {
        $curlData = [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER => $this->getAuthHeaders(),
            CURLOPT_HEADER => 0,
            CURLOPT_FAILONERROR => true,
            CURLOPT_URL => $url
        ];

        if (is_array($data)) {
            $curlData = $curlData + $data;
        }

        curl_setopt_array($handle, $curlData);

        return RequestHeader::create($curlData);
    }

    /**
     * @param RequestHeader $headers
     */
    public function setLastRequestHeaders(RequestHeader $headers)
    {
        $this->lastRequestHeaders = $headers;
    }

    /**
     * Get Authentication Headers
     * @return mixed
     * @throws MarelloException
     */
    private function getAuthHeaders()
    {
        if ((!isset($this->credentials['username']) && empty($this->credentials['username']))
            || (!isset($this->credentials['api_key']) && empty($this->credentials['api_key']))) {
            throw new MarelloException(
                'No credentials provided. Please call Client::setAuth() before making requests.'
            );
        }

        $auth = new Authentication($this->credentials['username'], $this->credentials['api_key']);

        return $auth->getHeaders();
    }

    /**
     * Get request headers from last cURL session
     * @return int
     */
    public function getLastRequestHeaders()
    {
        return $this->lastRequestHeaders;
    }

    /**
     * Get last response from last cURL session
     * @return Response
     */
    public function getLastResponse()
    {
        return $this->lastResponse;
    }

    /**
     * Returns if call made has produced an error
     * @return boolean|null
     */
    public function hasError()
    {
        return ($this->lastResponse->getData('error'));
    }

    /**
     * Prepare get params for request
     * array('paramname' => 'paramvalue')
     * => <base_uri>/<path>/paramvalue
     * @param string $url
     * @param array $params
     * @return string $uri
     */
    private function prepareGetParams($url, array $params)
    {
        if (empty($params)) {
            return $url;
        }

        if (array_key_exists('id', $params)) {
            $uri = $this->prepareRestUri($params['id'], $url);
        } else {
            $uriParams = sprintf('filter?%s', http_build_query($params));
            $uri = $this->prepareRestUri($uriParams, $url);
        }

        return $uri;
    }

    /**
     * REST GET call
     * @param string $path
     * @param array $query
     * @return mixed
     */
    public function restGet($path, array $query = [])
    {
        $url = $this->prepareRestUri($path);
        // add query params to url
        $fullUrl = $this->prepareGetParams($url, $query);

        return $this->callServer($fullUrl, null, $query);
    }

    /**
     * REST POST call
     * @param $path
     * @param array $query
     * @return mixed
     * @throws MarelloException
     */
    public function restPost($path, array $query = [])
    {
        if (empty($query)) {
            throw new MarelloException('Query cannot be empty when a POST call is made.');
        }

        $url = $this->prepareRestUri($path);
        $curlOptions = [
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => json_encode($query)
        ];

        return $this->callServer($url, $curlOptions, $query);
    }

    /**
     * REST PUT call
     * @param $path
     * @param array|null $query
     * @return mixed
     * @throws MarelloException
     */
    public function restPut($path, array $query = [])
    {
        if (empty($query)) {
            throw new MarelloException('Query cannot be empty when a PUT call is made.');
        }

        $url = $this->prepareRestUri($path);
        $curlOptions = [
            CURLOPT_CUSTOMREQUEST => 'PUT',
            CURLOPT_POSTFIELDS => json_encode($query)
        ];

        return $this->callServer($url, $curlOptions, $query);
    }

    /**
     * REST DELETE call
     * @param $path
     * @param array $query
     * @return mixed
     * @throws MarelloException
     */
    public function restDelete($path, array $query = [])
    {
        if (empty($query)) {
            throw new MarelloException('Query cannot be empty when a DELETE call is made.');
        }

        $url = $this->prepareRestUri($path);
        $curlOptions = [
            CURLOPT_CUSTOMREQUEST => 'DELETE',
            CURLOPT_POSTFIELDS => json_encode($query)
        ];

        return $this->callServer($url, $curlOptions, $query);
    }
}
